File: output/cpp/tests/08_hsm.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>
#include <vector>
#include <algorithm>

class HSMForwardFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    HSMForwardFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class HSMForwardFrameContext {
public:
    HSMForwardFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    HSMForwardFrameContext(HSMForwardFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class HSMForwardCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<HSMForwardFrameEvent> forward_event;
    std::unique_ptr<HSMForwardCompartment> parent_compartment;

    explicit HSMForwardCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<HSMForwardCompartment> clone() const {
        auto c = std::make_unique<HSMForwardCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class HSMForward {
private:
    std::vector<std::unique_ptr<HSMForwardCompartment>> _state_stack;
    std::unique_ptr<HSMForwardCompartment> __compartment;
    std::unique_ptr<HSMForwardCompartment> __next_compartment;
    std::vector<HSMForwardFrameContext> _context_stack;

    void __kernel(HSMForwardFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            HSMForwardFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                HSMForwardFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    HSMForwardFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(HSMForwardFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Child") {
            _state_Child(__e);
        } else if (state_name == "Parent") {
            _state_Parent(__e);
        }
    }

    void __transition(std::unique_ptr<HSMForwardCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Child(HSMForwardFrameEvent& __e) {
        if (__e._message == "event_a") {
            event_log.push_back("Child:event_a");
        } else if (__e._message == "event_b") {
            event_log.push_back("Child:event_b_forward");
            _state_Parent(__e);
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        }
    }

    void _state_Parent(HSMForwardFrameEvent& __e) {
        if (__e._message == "event_a") {
            event_log.push_back("Parent:event_a");
        } else if (__e._message == "event_b") {
            event_log.push_back("Parent:event_b");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        }
    }

public:
    std::vector<std::string> event_log = {};

    HSMForward() {
        // HSM: Create parent compartment chain
        auto __parent_comp_0 = std::make_unique<HSMForwardCompartment>("Parent");
        __compartment = std::make_unique<HSMForwardCompartment>("Child");
        __compartment->parent_compartment = std::move(__parent_comp_0);
        HSMForwardFrameEvent __frame_event("$>");
        HSMForwardFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void event_a() {
        HSMForwardFrameEvent __e("event_a");
        HSMForwardFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void event_b() {
        HSMForwardFrameEvent __e("event_b");
        HSMForwardFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::vector<std::string> get_log() {
        HSMForwardFrameEvent __e("get_log");
        HSMForwardFrameContext __ctx(std::move(__e), std::any(std::vector<std::string>()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::vector<std::string>>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 08: HSM Forward ===\n");
    HSMForward s;

    s.event_a();
    auto log = s.get_log();
    if (std::find(log.begin(), log.end(), "Child:event_a") == log.end()) {
        printf("FAIL: Expected 'Child:event_a' in log\n");
        assert(false);
    }
    printf("After event_a: Child:event_a found\n");

    s.event_b();
    log = s.get_log();
    if (std::find(log.begin(), log.end(), "Child:event_b_forward") == log.end()) {
        printf("FAIL: Expected 'Child:event_b_forward' in log\n");
        assert(false);
    }
    if (std::find(log.begin(), log.end(), "Parent:event_b") == log.end()) {
        printf("FAIL: Expected 'Parent:event_b' in log (forwarded)\n");
        assert(false);
    }
    printf("After event_b (forwarded): Child:event_b_forward and Parent:event_b found\n");

    printf("PASS: HSM forward works correctly\n");
    return 0;
}

File: output/go/tests/08_hsm.go

package main

import "fmt"
import "os"

type HSMForwardFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type HSMForwardFrameContext struct {
    _event  HSMForwardFrameEvent
    _return any
    _data   map[string]any
}

type HSMForwardCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *HSMForwardFrameEvent
    parentCompartment *HSMForwardCompartment
}

func newHSMForwardCompartment(state string) *HSMForwardCompartment {
    return &HSMForwardCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *HSMForwardCompartment) copy() *HSMForwardCompartment {
    nc := &HSMForwardCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type HSMForward struct {
    _state_stack []*HSMForwardCompartment
    __compartment *HSMForwardCompartment
    __next_compartment *HSMForwardCompartment
    _context_stack []HSMForwardFrameContext
    log []string
}

func NewHSMForward() *HSMForward {
    s := &HSMForward{}
    s._state_stack = make([]*HSMForwardCompartment, 0)
    s._context_stack = make([]HSMForwardFrameContext, 0)
    // HSM: Create parent compartment chain
    __parent_comp_0 := newHSMForwardCompartment("Parent")
    __parent_comp_0.parentCompartment = nil
    s.__compartment = newHSMForwardCompartment("Child")
    s.__compartment.parentCompartment = __parent_comp_0
    s.__next_compartment = nil
    __frame_event := HSMForwardFrameEvent{_message: "$>", _parameters: nil}
    __ctx := HSMForwardFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *HSMForward) __kernel(__e *HSMForwardFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &HSMForwardFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &HSMForwardFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &HSMForwardFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *HSMForward) __router(__e *HSMForwardFrameEvent) {
    switch s.__compartment.state {
    case "Child":
        s._state_Child(__e)
    case "Parent":
        s._state_Parent(__e)
    }
}

func (s *HSMForward) __transition(next *HSMForwardCompartment) {
    s.__next_compartment = next
}

func (s *HSMForward) EventA() {
    __e := HSMForwardFrameEvent{_message: "EventA"}
    __ctx := HSMForwardFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMForward) EventB() {
    __e := HSMForwardFrameEvent{_message: "EventB"}
    __ctx := HSMForwardFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMForward) GetLog() []string {
    __e := HSMForwardFrameEvent{_message: "GetLog"}
    __ctx := HSMForwardFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result []string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.([]string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMForward) _state_Parent(__e *HSMForwardFrameEvent) {
    if __e._message == "EventA" {
        s.log = append(s.log, "Parent:event_a")
    } else if __e._message == "EventB" {
        s.log = append(s.log, "Parent:event_b")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    }
}

func (s *HSMForward) _state_Child(__e *HSMForwardFrameEvent) {
    if __e._message == "EventA" {
        s.log = append(s.log, "Child:event_a")
    } else if __e._message == "EventB" {
        s.log = append(s.log, "Child:event_b_forward")
        s._state_Parent(__e)
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    }
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("=== Test 08: HSM Forward ===")
	sm := NewHSMForward()

	// event_a should be handled by Child (no forward)
	sm.EventA()
	log := sm.GetLog()
	if !contains(log, "Child:event_a") {
		fmt.Printf("FAIL: Expected 'Child:event_a' in log, got %v\n", log)
		os.Exit(1)
	}
	fmt.Printf("After EventA: %v\n", log)

	// event_b should forward to Parent
	sm.EventB()
	log = sm.GetLog()
	if !contains(log, "Child:event_b_forward") {
		fmt.Printf("FAIL: Expected 'Child:event_b_forward' in log, got %v\n", log)
		os.Exit(1)
	}
	if !contains(log, "Parent:event_b") {
		fmt.Printf("FAIL: Expected 'Parent:event_b' in log (forwarded), got %v\n", log)
		os.Exit(1)
	}
	fmt.Printf("After EventB (forwarded): %v\n", log)

	fmt.Println("PASS: HSM forward works correctly")
}
